### ISSUE TITLE
Sporadic hanging of the client

### ISSUE THREAD

**Issue by huntc** [landlord/landlord#44]
Upon following the instructions in the README, the client appears to hang when invoking multiple times (not always though):

```
$ echo Testing... | landlord/target/release/landlord -cp landlordd/test/target/scala-2.12/classes "-Dgreeting=Welcome" example.Hello ArgOne ArgTwo
Argument #1: ArgOne
Argument #2: ArgTwo
Welcome
Testing...
Good Bye!
$ 
$ echo Testing... | landlord/target/release/landlord -cp landlordd/test/target/scala-2.12/classes "-Dgreeting=Welcome" example.Hello ArgOne ArgTwo
```

At this point, I'm unable to CTRL-C from the process. I have to SIGKILL it from elsewhere.

**Comment by huntc** [comment]
I don't seem to have the same problem with the count program though...

**Comment by longshorej** [comment]
Sounds like a bug in `JvmExecutor` - likely some threads still running. We should probably print what's running to debug.

Trying to get macOS back onto an old macMini of mine so I can reproduce.

**Comment by longshorej** [comment]
Could you attach landlordd's output when this happens @huntc? I've been running this command in a loop for an hour without fail, although that's on linux with an 8 thread / 32gb mem machine..

**Comment by huntc** [comment]
There was no output in Landlord of note.

May be try running it in a Docker container (my PR updates the readme to show how). You can then limit CPU and play around with timings. Wdyt?

**Comment by longshorej** [comment]
No such luck yet - running both server and client in different containers, playing with shares, etc. Even tried on an old Core2 E8500 without luck.

When it hangs, it looks like you're not getting any output, is that right?

**Comment by huntc** [comment]
Yeah. No output. I’ll look further into this if you’d like. May be it is OS X specific.

**Comment by huntc** [comment]
It seems to be related to the echo of stdin... Perhaps some race condition there? If manually CTRL-Z when it "hangs" then it will actually always exit.

**Comment by huntc** [comment]
There are other times though when it doesn't even output the args and then hangs up...

**Comment by huntc** [comment]
More suspicion around the Rust client. Here's my latest:

First, I'm running socat to intercept and report on UDS traffic:

```
sudo socat -t100 -x -v \
  UNIX-LISTEN:/tmp/duplicatesocket,mode=777,reuseaddr,fork \
  UNIX-CONNECT:/var/run/landlord/landlordd.sock
```

...then run `landlord` on the duplicate socket:

```
$ echo "hi" | landlord/target/release/landlord -H unix:///tmp/duplicatesocket -cp landlordd/test/target/scala-2.12/classes  example.Hello ArgOne ArgTwo
```

After a few seconds I see:

```
landlord: read_hadler crashed, Custom { kind: UnexpectedEof, error: StringError("failed to fill whole buffer") }
```

...which looks suspicious. The bytes appear to be delivered correctly over the UDS:

```
< 2018/06/29 14:21:16.633422  length=4 from=0 to=3
 00 00 00 08                                      ....
--
> 2018/06/29 14:21:16.635603  length=3 from=10275 to=10277
 68 69 0a                                         hi.
```

That said, I'm not seeing any output from the JvmExecutor, although debugging the Hello app shows that it writes to the output buffer... So perhaps there's a problem on both sides.

**Comment by huntc** [comment]
Note, in the above, I see the JvmExecutor having sent the exit code, so it appears as though the JvmExecutor is behaving correctly... Perhaps UDS after all...

**Comment by huntc** [comment]
I'm thinking the client again. Here's another example:

```
< 2018/06/29 14:54:42.293531  length=4 from=0 to=3
 00 00 00 28                                      ...(
--
< 2018/06/29 14:54:42.297657  length=102 from=4 to=105
 6f 00 00 00 13 41 72 67 75 6d 65 6e 74 20 23 31  o....Argument #1
 3a 20 41 72 67 4f 6e 65 6f 00 00 00 01 0a        : ArgOneo.....
 6f 00 00 00 13 41 72 67 75 6d 65 6e 74 20 23 32  o....Argument #2
 3a 20 41 72 67 54 77 6f 6f 00 00 00 01 0a        : ArgTwoo.....
 6f 00 00 00 1f 57 65 6c 63 6f 6d 65 2e 20 50 6c  o....Welcome. Pl
 65 61 73 65 20 74 79 70 65 20 61 20 6d 65 73 73  ease type a mess
 61 67 65 2e 6f 00 00 00 01 0a                    age.o.....
--
> 2018/06/29 14:54:42.299241  length=3 from=10275 to=10277
 68 69 0a                                         hi.
```

In this example, the daemon has sent everything with stdin being received after... Meanwhile, the client, after a few seconds:

```$ echo "hi" | landlord/target/release/landlord -H unix:///tmp/duplicatesocket -cp landlordd/test/target/scala-2.12/classes  example.Hello ArgOne ArgTwo
Argument #1: ArgOne
Argument #2: ArgTwo
Welcome. Please type a message.
landlord: read_hadler crashed, Custom { kind: UnexpectedEof, error: StringError("failed to fill whole buffer") }
```

**Comment by longshorej** [comment]
I suspect this is fixed with Alpakka 0.20. Will leave the issue open until we can confirm.

**Comment by longshorej** [comment]
This was an issue with UNIX domain sockets, not landlord/landlordd in particular.
